File: Week_02/G20200343030632/Leetcode_144_632.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        # return self.preorderTraversal_v1(root)
        return self.preorderTraversal_v2(root)

    def preorderTraversal_v1(self, root: TreeNode) -> List[int]:
        res = []
        def helper(root: TreeNode, res: List[int]):
            if root:
                res.append(root.val)
                helper(root.left, res)
                helper(root.right, res)
        helper(root, res)
        return res

    def preorderTraversal_v2(self, root: TreeNode) -> List[int]:
        """
            迭代做法, 利用栈.
            前序遍历的进栈顺序是先右, 左, 根
        """
        stack, res = [], []
        white, gray = 0, 1
        stack.append((white, root))
        while stack:
            color, node = stack.pop()
            if node is None:
                continue
            if color == white:
                stack.append((white, node.right))
                stack.append((white, node.left))
                stack.append((gray, node))
            else:
                res.append(node.val)
        return res







# @lc code=end

File: Week_02/G20200343030632/Leetcode_144_632.java
import java.util.Deque;
import java.util.LinkedList;



/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        // return preorderTraversal_v1(root);
        return preorderTraversal_v2(root);
    }

    public List<Integer> preorderTraversal_v1(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        helper(root, res);
        return res;
    }
    private void helper(TreeNode cur, List<Integer> res) {
        if (cur != null) {
            res.add(cur.val);
            helper(cur.left, res);
            helper(cur.right, res);
        }
    }

    public List<Integer> preorderTraversal_v2(TreeNode root) {
        // 迭代
        List<Integer> res = new LinkedList<>();
        Deque<TreeNode> stack = new LinkedList<>();
        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                res.add(root.val);
                stack.addFirst(root);
                root = root.left;
            }
            root = stack.removeFirst();
            root = root.right;
        }
        return res;

    }
}